Handle `drawer_on_selection` UI

carlosmuvi-stripe opened the pull request:
# Summary
- Adds basic support for SDUX needed for NME
- Adds a new entry to `ModalBottomSheet` to render the `GenericScreen` content

# Motivation
:notebook_with_decorative_cover: &nbsp;**Handle `drawer_on_selection` UI**
:globe_with_meridians: &nbsp;[BANKCON-11425](https://jira.corp.stripe.com/browse/BANKCON-11425)
<!-- Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue. -->

# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [x] Added screenshot tests
- [x] Manually verified

Test it yourself: Networking merchant, testmode, use `kalospilesooo@gmail.com` as RUX, Link account picker, select the disabled account to see the modal. 

# Screenshots

<img width="326" alt="image" src="https://github.com/stripe/stripe-android/assets/99293320/c934c37e-f832-41fa-b196-a07e869ee56f">


# Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

carlosmuvi-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/common/ModalBottomSheetContent.kt:
```diff
@@ -119,6 +122,20 @@ internal fun LegalDetailsBottomSheetContent(
+    GenericScreen(
+        state = GenericScreenState(generic, inModal = true),
+        onPrimaryButtonClick = onConfirmModalClick,
+        onSecondaryButtonClick = { TODO("Secondary button click not yet implemented") },
```
this drawer does not have a secondary button, but supporting this on a follow up PR.

carlosmuvi-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/common/ShapedIcon.kt:
```diff
@@ -29,14 +28,19 @@ private val iconSize = 20.dp
 internal fun ShapedIcon(
     painter: Painter,
+    modifier: Modifier = Modifier,
+    iconSize: IconSize = IconSize.Medium,
```
Adapted our icon to support sizes, matching web - https://git.corp.stripe.com/stripe-internal/stripe-js-v3/blob/5573b9b445dfaac104225d6e7969035f3cc26701/src/linkedAccounts/shared-ui/v3/Icon/index.tsx#L47

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/common/ModalBottomSheetContent.kt:
```diff
@@ -119,6 +122,20 @@ internal fun LegalDetailsBottomSheetContent(
 
+@Composable
+internal fun GenericBottomSheetContent(
+    generic: FinancialConnectionsGenericInfoScreen,
```
```suggestion
    screen: FinancialConnectionsGenericInfoScreen,
```

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+    var footerHeight by remember { mutableStateOf(0.dp) }
+    var contentHeight by remember { mutableStateOf(0.dp) }
+
+    val spacing by remember {
+        derivedStateOf {
+            ((containerHeight - footerHeight - contentHeight) / 2).takeIf {
+                state.screen.options?.verticalAlignment == VerticalAlignment.Centered && it > 0.dp
+            }
+        }
+    }
```
Do you think we should move this into `Layout` and then have an argument `centerVerticallyIfEnoughSpace` or something like that?

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+            {
+                GenericFooter(
+                    payload = it,
+                    modifier = Modifier.onGloballyPositioned {
+                        with(density) {
+                            footerHeight = it.size.height.toDp()
+                        }
+                    },
```
Since this is used a few times: Can we create a Modifier extension method here, like `Modifier.onHeightDetermined(action: (Dp) -> Unit)`?

carlosmuvi-stripe replied:
removed centering logic for this PR, let's revisit it as soon as needed!

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+                )
+            }
+
+            Spacer(modifier = Modifier.height(20.dp))
```
Can we use `Arrangment.spacedBy()` for this? This way, it’ll only be rendered if both a header and a body are present.

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+                }
+            }
+
+            if (index != payload.entries.lastIndex) {
+                Spacer(modifier = Modifier.size(24.dp))
+            }
```
Can this use `Arrangement.spacedBy()` instead?

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+        }
+
+        if (payload.subtitle != null) {
+            Spacer(modifier = Modifier.height(16.dp))
```
I assume we’ll always have a title if we have a subtitle? Otherwise, we could also use `Arrangement.spacedBy()` here for consistent spacing.

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+        }
+
+        payload.secondaryCta?.let { secondaryCta ->
+            Spacer(modifier = Modifier.size(8.dp))
```
Same here about consistent spacing.

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+}
+
+@Composable
+private fun GenericCTA(
```
```suggestion
private fun GenericButton(
```

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/generic/GenericScreen.kt:
```diff
@@ -0,0 +1,345 @@
+                url = it,
+                contentDescription = null,
+                imageLoader = LocalImageLoader.current,
+                errorContent = { },
```
```suggestion
```

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/features/linkaccountpicker/LinkAccountPickerViewModel.kt:
```diff
@@ -284,8 +285,10 @@ internal class LinkAccountPickerViewModel @AssistedInject constructor(
         if (drawerOnSelection != null) {
-            // TODO@carlosmuvi handle drawer display.
-            logger.debug("Drawer on selection: $drawerOnSelection")
+            presentNoticeSheet(
```
We could rename this to `presentSheet`.

I also imagine that we can get rid of the subclasses and just have a single `presentSheet(SheetContent)` method, by using the generic screen data class for the existing `Legal` and `DataAccess` subclasses. What do you think?

Not something that needs to happen in this pull request. I’d be happy to do it as a follow-up!

carlosmuvi-stripe replied:
Agreed! Doing the renaming for now.

carlosmuvi-stripe replied:
Renamed it as part of https://github.com/stripe/stripe-android/commit/fc614e6b9a094515745be42cf3e32b60bae9dd64

carlosmuvi-stripe commented:
PR feedback - https://github.com/stripe/stripe-android/pull/8743/commits/fc614e6b9a094515745be42cf3e32b60bae9dd64

- Deleted centering logic
- Used consistent spacing via arrangement across the screen
- Renamed present notice sheet.

carlosmuvi-stripe commented:
# PR stack

This pull request is part of a PR stack:

_No data yet..._

------

This comment is managed by [#work-cli](https://stripe.slack.com/archives/C02EN462DML)
_UUID: 6915807b-e91a-454d-9034-d1f5c1c1155e_

* To request reviews on the entire stack, use `work stack review`
* To merge your PRs, start with the first (oldest) PR on the stack
* For more help, `work help stack`
